File: trunk/ShadowFrameworkLite20/src/shadow/renderer/data/SFStructureArrayData.java
package shadow.renderer.data;

import shadow.math.SFValuenf;
import shadow.pipeline.SFPipeline;
import shadow.pipeline.SFPipelineStructure;
import shadow.pipeline.SFStructureArray;
import shadow.pipeline.SFStructureData;
import shadow.system.data.SFDataset;
import shadow.system.data.SFNamedParametersObject;
import shadow.system.data.objects.SFBinaryVertexArrayList;
import shadow.system.data.objects.SFGenericFixedFloat;
import shadow.system.data.objects.SFString;

public class SFStructureArrayData<T extends SFGenericFixedFloat> extends SFDataAsset<SFStructureArray> implements SFDataset{

	private SFString structure=new SFString();
	private SFStructureArray array;
	private SFBinaryVertexArrayList<T> values;
	
	public SFStructureArrayData(T t){
		values=new SFBinaryVertexArrayList<T>(t);
		SFNamedParametersObject parameters=new SFNamedParametersObject();
		//parameters.addObject("vertexSize", this.vertexSize);
		parameters.addObject("structure", structure);
		parameters.addObject("values", values);
		setData(parameters);
	}
	
	public void setStructure(String structureName){
		this.structure.setString(structureName);
	}
	
	public SFStructureArray getArray() {
		if(array==null){
			String structureName=structure.getString();
			SFPipelineStructure materialStructureInstance=((SFPipelineStructure)(SFPipeline.getModule(structureName)));
			array=SFPipeline.getSfPipelineMemory().generateStructureData(materialStructureInstance); 
			int n=values.getSize();
			SFStructureData data=new SFStructureData(materialStructureInstance);
			for (int i = 0; i < n; i++) {
				SFValuenf[] value=data.getValues();
				for (int j = 0; j < value.length; j++) {
					values.getValue(i, j, value[j]);
				}
				array.generateElement();
				array.setElement(i, data);
			}
		}
		return array;
	}

	public void setArray(SFStructureArray array) {
		this.array = array;
		values.getDataObject().clear();
		SFStructureData data=new SFStructureData(array.getPipelineStructure());
		for (int i = 0; i < array.getElementsCount(); i++) {
			array.getElement(i, data);
			SFValuenf[] value=data.getValues();
			values.addValue(value);
		}
	}
	
	@Override
	protected SFStructureArray buildResource() {
		return getArray();
	}

	
	public void invalidate() {
	}

}

File: trunk/ShadowFrameworkJS/src/shadow/renderer/data/SFStructureArrayData.js


function SFStructureArrayData(t){
	this.t=t;
	this.values=new SFBinaryVertexArrayList(t);
	this.structure=new SFString();
	var parameters=new SFNamedParametersObject();
	parameters.addObject(this.structure);
	parameters.addObject(this.values);
	this.setData(parameters);
}

inherit(SFStructureArrayData,SFDataAsset);


SFDataAsset.prototype["setStructure"]=function(structureName){
			this.structure.setString(structureName);
		};


SFStructureArrayData.prototype["generateNewDatasetInstance"]=function(){
	var array=new SFStructureArrayData(this.t);
	return array;
};

SFStructureArrayData.prototype["getArray"]=function(){
			if(this.array===undefined){
				var structureName=this.structure.getString();
				var materialStructureInstance=SFPipeline_getModule(structureName);
				this.array=SFPipeline_getSfPipelineMemory().generateStructureData(materialStructureInstance); 
				var n=this.values.getSize();
				var data=new SFStructureData(materialStructureInstance);
				for (var i = 0; i < n; i++) {
					var value=data.getValues();
					for (var j = 0; j < value.length; j++) {
						this.values.getValue(i, j, value[j]);
					}
					this.array.generateElement();
					this.array.setElement(i, data);
				}
			}
			return this.array;
		};

SFStructureArrayData.prototype["setArray"]=function(array){
		this.array = array;
		this.values.getDataObject().clear();
		var data=new SFStructureData(array.getPipelineStructure());
		for (var i = 0; i < array.getElementsCount(); i++) {
			array.getElement(i, data);
			var value=data.getValues();
			values.addValue(value);
		}
	};


SFStructureArrayData.prototype["buildResource"]=function(){
		return this.getArray();
	};

SFStructureArrayData.prototype["invalidate"]=function(){
		//Nothing to do.
	};

